nuxeo-connect-client/src/main/java/org/nuxeo/connect/packages/PackageManagerImpl.java

```diff
@@ -1096,24 +1103,48 @@ public void checkOptionalDependenciesOnInstalledPackages(DependencyResolution re
+                    // find matching installed version
+                    List<Version> installedVersions = findLocalPackageInstalledVersions(pkgOptDep.getName());
+                    boolean hasAnInstalledMatch = false;
+                    for (Version version : installedVersions) {
+                        if (pkgOptDep.getVersionRange().matchVersion(version)) {
+                            hasAnInstalledMatch = true;
+                            break;
+                        }
+                    }
+                    if (!hasAnInstalledMatch) {
+                        // if optional dependency is not already installed, check if it is currently being installed
                         for (String pkgId : res.getOrderedPackageIdsToInstall()) {
                             DownloadablePackage pkgToInstall = getPackage(pkgId);
                             if (matchDependency(pkgOptDep, pkgToInstall)) {
-                                log.info(String.format(
+                                res.logInfo(String.format(
                                         "As package '%s' has an optional dependency on package '%s' currently being installed, it will be reinstalled.",
                                         installedPkg, pkgToInstall));
                                 packagesToReinstall.add(installedPkg);
                             }
                         }
                     } else {
-                        for (String pkgId : res.getOrderedPackageIdsToRemove()) {
-                            DownloadablePackage pkgToRemove = getPackage(pkgId);
-                            if (matchDependency(pkgOptDep, pkgToRemove)) {
-                                log.info(String.format(
-                                        "As package '%s' has an optional dependency on package '%s' currently being uninstalled, it will be reinstalled.",
-                                        installedPkg, pkgToRemove));
-                                packagesToReinstall.add(installedPkg);
+                        // find matching currently installing version (upgrade)
+                        boolean hasAnUpgradingMatch = false;
+                        for (String pkgId : res.getOrderedPackageIdsToInstall()) {
+                            DownloadablePackage pkgToInstall = getPackage(pkgId);
+                            if (pkgToInstall.getName().equals(pkgOptDep.getName())
+                                    && pkgOptDep.getVersionRange().matchVersion(pkgToInstall.getVersion())) {
+                                hasAnUpgradingMatch = true;
+                                break;
+                            }
+                        }
+                        if (!hasAnUpgradingMatch) {
+                            // if optional dependency is already installed and is not currently being upgraded, check if
+                            // it is currently being uninstalled
```

Comment:
Is this rewrite true?
```// if 'installedPkg''s optional dependency ('pkgOptDep') is already installed and is not currently being upgraded, then reinstall 'installedPkg' if 'pkgOptDep' is being uninstalled```
You should reuse the variable names and be more precise in order to explain this algorithm which becomes complex.

Thus, I'm wondering if this cannot be simplified, at least in the writing, optionally in the algorithm.
For instance, instead of this for loop 
```for (String pkgId : res.getOrderedPackageIdsToRemove()) {
DownloadablePackage pkgToRemove = getPackage(pkgId);```
don't we rather need something as simple as `if (containsDependency(pkgOptDep, res.getRemovePackageIds())) ...`?